### PR TITLE
Remove two extra lexer passes

### PR DESCRIPTION
**Pull request by sand1k**
Remove preparser lexer pass (searches for variable declarations inside a scope) and pass for searching of "eval" and "arguments" literals.

### CODE REVIEW

File: jerry-core/parser/js/scopes-tree.cpp [hunk 1]
```diff
@@ -519,9 +580,8 @@ scopes_tree_count_literals_in_blocks (scopes_tree tree)
   bool header = true;
   for (instr_pos = 0; instr_pos < tree->instrs_num; instr_pos++)
   {
-    op_meta *om = extract_op_meta (tree, instr_pos);
-    if (om->op.op_idx != VM_OP_VAR_DECL
-        && om->op.op_idx != VM_OP_META && !header)
+    op_meta *om = extract_op_meta (tree->instrs, instr_pos);
```
Comment (egavrin) [inline jerry-core/parser/js/scopes-tree.cpp]: `om_p` :)

File: jerry-core/parser/js/scopes-tree.cpp [hunk 2]
```diff
@@ -529,36 +589,49 @@ scopes_tree_count_literals_in_blocks (scopes_tree tree)
...
+
   for (uint8_t child_id = 0; child_id < tree->t.children_num; child_id++)
   {
     result += scopes_tree_count_literals_in_blocks (*(scopes_tree *) linked_list_element (tree->t.children, child_id));
   }
+
   for (; instr_pos < tree->instrs_num; instr_pos++)
   {
-    result += count_new_literals_in_instr (tree, instr_pos);
+    op_meta *om = extract_op_meta (tree->instrs, instr_pos);
```
Comment (egavrin) [inline jerry-core/parser/js/scopes-tree.cpp]: likewise

File: jerry-core/parser/js/scopes-tree.cpp [hunk 3]
```diff
@@ -529,36 +589,49 @@ scopes_tree_count_literals_in_blocks (scopes_tree tree)
...
+ *  then we shall recursively dump function declaration,
+ *  and finally, other instructions.
+ *
+ *  For each instructions block (size of block is defined in bytecode-data.h)
+ *  literal indexes 'hash' table is filled.
+ */
 static void
-merge_subscopes (scopes_tree tree, vm_instr_t *data, lit_id_hash_table *lit_ids)
+merge_subscopes (scopes_tree tree, /**< scopes tree to merge */
+                 vm_instr_t *data, /**< instruction array, where the scopes are merged to */
```
Comment (egavrin) [inline jerry-core/parser/js/scopes-tree.cpp]: likewise

File: jerry-core/parser/js/scopes-tree.cpp [hunk 4]
```diff
@@ -529,36 +589,49 @@ scopes_tree_count_literals_in_blocks (scopes_tree tree)
     {
       header = false;
     }
-    result += count_new_literals_in_instr (tree, instr_pos);
+    result += count_new_literals_in_instr (om);
   }
+
+  for (vm_instr_counter_t var_decl_pos = 0; var_decl_pos < tree->var_decls_num; var_decl_pos++)
+  {
+    op_meta *om = extract_op_meta (tree->var_decls, var_decl_pos);
```
Comment (egavrin) [inline jerry-core/parser/js/scopes-tree.cpp]: likewise

File: jerry-core/parser/js/scopes-tree.h [hunk 5]
```diff
@@ -39,12 +39,19 @@ typedef struct tree_header
...
+ */
 typedef struct
 {
-  tree_header t;
-  linked_list instrs;
-  vm_instr_counter_t instrs_num;
-  unsigned strict_mode:1;
+  tree_header t; /**< header */
+  linked_list instrs; /**< instructions */
+  vm_instr_counter_t instrs_num; /**< count of instructions */
```
Comment (egavrin) [inline jerry-core/parser/js/scopes-tree.h]: `instr_count`?

**Comment by sand1k** [comment]
@egavrin, fixed.

**Comment by egavrin** [comment]
Good to me

File: jerry-core/parser/js/scopes-tree.cpp [hunk 6]
```diff
@@ -638,6 +719,26 @@ scopes_tree_set_strict_mode (scopes_tree tree, bool strict_mode)
...
 }
 
+/**
+ * Set up a flag, indicating that "arguments" is used inside a scope
+ */
+void
+scopes_tree_set_arguments_used (scopes_tree tree) /**< scope */
+{
+  assert_tree (tree);
+  tree->ref_arguments = 1;
```
Comment (zherczeg) [inline jerry-core/parser/js/scopes-tree.cpp]: Shouldn't these be bools?

**Comment by zherczeg** [comment]
LGTM

File: jerry-core/parser/js/scopes-tree.cpp (hunk 6 above)
Comment (sand1k) [inline jerry-core/parser/js/scopes-tree.cpp]: Fixed.

**Comment by egavrin** [comment]
`make push`

**Pull request merged**